ledger/participant-state/kvutils/app/src/main/scala/com/daml/ledger/participant/state/kvutils/app/Runner.scala

```diff
@@ -50,11 +51,21 @@ class Runner[T <: KeyValueLedger, Extra](name: String, factory: LedgerFactory[T,
         _ <- AkkaResourceOwner.forActorSystem(() => system)
         _ <- AkkaResourceOwner.forMaterializer(() => materializer)
 
-        readerWriter <- factory.owner(config.ledgerId, config.participantId, config.extra)
-        ledger = new KeyValueParticipantState(readerWriter, readerWriter)
-        _ <- ResourceOwner.forFuture(() =>
-          Future.sequence(config.archiveFiles.map(uploadDar(_, ledger))))
-        _ <- startParticipant(config, ledger)
+        // initialize all configured participants
+        _ <- config.participants.foldLeft(ResourceOwner.successful(())) {
+          case (acc, participantConfig) =>
+            for {
+              _ <- acc
+              readerWriter <- factory.owner(
+                config.ledgerId,
+                participantConfig.participantId,
+                config.extra)
+              ledger = new KeyValueParticipantState(readerWriter, readerWriter)
+              _ <- ResourceOwner.forFuture(() =>
+                Future.sequence(config.archiveFiles.map(uploadDar(_, ledger))))
+              _ <- startParticipant(config, participantConfig, ledger)
+            } yield ()
+        }
```

Comment:
Can you use `ResourceOwner.sequence` here?

Reply:
That's much better of course. I honestly didn't think to look for it in `ResourceOwner`, but I actually did think "if only we had Applicative for resource owner here for sequence...".